### PR TITLE
PHP 8.0 and 8.1

### PR DESCRIPTION
**Pull request by Tybaze**
# PHP 8.0 & 8.1 compatibility

[No Enhancement - please considere to use Doctrine2 for your new Project]

Target : Maintains maximum backward compatibility down to 5.3 for the moment

PHP 8.0 and 8.1 brings a lot of incompatibility.
https://www.php.net/manual/en/migration80.incompatible.php
https://www.php.net/manual/en/migration81.incompatible.php

## Tests

I have to update travis, to include new php versions, and to remove php5.3 on Precise (letsencrypt root certificate is broken, impossible to download composer anymore)

I tested localy on PHP 5.3 and passed the build successfully.

alt text
[Travis Build Link](https://api.travis-ci.com/Tybaze/doctrine1.svg?branch=compat_php8.1)

| PHP version  |  Status |
| ------------- | --- |
| php5.3  |  🟢 |
| php5.4  |  🟢 |
| php5.5  |  🟢 |
| php5.6  |  🟢 |
| php7.0  |  🟢 |
| php7.1  |  🟢 |
| php7.2  |  🟢 |
| php7.3  |  🟢 |
| php7.4  |  🟢 |
| php8.0  |  🟢 |
| php8.1  |  🟢 |

## Future

One Day PHP 9 will broke compatibility with php < 7.4 due to the lacks of "mixed" keywords implementations.

Please considere rewriting asap your project to another framework, like Doctrine2 if your app cannot not support at least PHP 7.4 within a few months/years

### CODE REVIEW

**Comment by alquerci** [comment]
Hello @Tybaze 

Thanks for your contribution for the huge goal of PHP 8 support.

Did you execute tests based on the fix of tests execution, see #56 ?

**Comment by Tybaze** [comment]
> Did you execute tests based on the fix of tests execution, see #56 ?

No sorry didn't see it, I run test directly with travis. 
I'll try them and #86 on monday.

**Comment by Tybaze** [comment]
I was using a wrong way to run test  this PR 

I have reviewed #56 and patched my branch localy to do the test.
I also use your new #86 with docker compose, great thanks @alquerci 

Now I have updated the first post of the PR, new status will be, once #56 will be merged

| PHP version  |  Status |
| ------------- | --- |
| php8.0  |  🟢 |
| php8.1  |  🔴 |

Working on PHP8.1...

**Comment by Tybaze** [comment]
Everything is working now on PHP8.0 and PHP8.1 👍 

How do I test : 
```
git clone git@github.com:Tybaze/doctrine1.git
git remote add alquerci https://github.com/alquerci/doctrine1.git
git fetch alquerci

# PR85
git checkout compat_php8.1

# PR86 
git merge alquerci/add-docker-compose

# PR56 
git merge alquerci/fix-tests-to-be-able-to-finish-it-without-a-fatal-error

# Fix a very small conflict in .gitignore
git commit 

# modify file ./tests/bin/test , replace variable "skipPHPVersions" with "php9"
# Then run the test
./tests/bin/test
```
And we get it :
| PHP version  |  Status |
| ------------- | --- |
| php5.3  |  🟢 |
| php5.4  |  🟢 |
| php5.5  |  🟢 |
| php5.6  |  🟢 |
| php7.0  |  🟢 |
| php7.1  |  🟢 |
| php7.2  |  🟢 |
| php7.3  |  🟢 |
| php7.4  |  🟢 |
| php8.0  |  🟢 |
| php8.1  |  🟢 |

File: lib/Doctrine/Connection/Profiler.php [hunk 1]
```diff
@@ -109,7 +109,7 @@ public function __call($m, $a)
      * @param mixed $key
      * @return Doctrine_Event
```
Comment (thePanz) [inline lib/Doctrine/Connection/Profiler.php:110]:
```suggestion
     * @return Doctrine_Event|null
```

File: lib/Doctrine/Connection/Profiler.php [hunk 2]
```diff
@@ -123,7 +123,7 @@ public function get($key)
      *
      * @return array        all events in an array
```
Comment (thePanz) [inline lib/Doctrine/Connection/Profiler.php:124]:
```suggestion
     * @return Doctrine_Event[] All events in an array
```

File: lib/Doctrine/Connection/Profiler.php [hunk 3]
```diff
@@ -154,7 +156,7 @@ public function count()
      *
      * @return Doctrine_Event
```
Comment (thePanz) [inline lib/Doctrine/Connection/Profiler.php:157]:
```suggestion
     * @return Doctrine_Event|null
```

File: lib/Doctrine/Connection/Sqlite.php [hunk 4]
```diff
@@ -87,8 +93,18 @@ public function connect()
             return false;
```
Comment (thePanz) [inline lib/Doctrine/Connection/Sqlite.php:93]: The method is annotated to return void, let's not return a boolean :)
```suggestion
            return;
```

File: lib/Doctrine/Record.php [hunk 5]
```diff
@@ -798,6 +798,33 @@ private function prepareIdentifiers($exists = true)
...
+     * @param string $serialized                Doctrine_Record as serialized string
+     * @throws Doctrine_Record_Exception        if the cleanData operation fails somehow
+     * @return void
+     */
+    public function unserialize($serialized)
+    {
+        $array = unserialize($serialized);
+
+        $this->__unserialize($array);
+
```
Comment (thePanz) [inline lib/Doctrine/Record.php:818]: Remove the extra line here

File: lib/Doctrine/Table.php [hunk 6]
```diff
@@ -2991,26 +3029,30 @@ public function serialize()
...
+        $this->_fieldNames = $data[4];
+        $this->_columnNames = $data[5];
+        $this->columnCount = $data[6];
+        $this->hasDefaultValues = $data[7];
+        $this->_options = $data[8];
+        $this->_invokedMethods = $data[9];
+        $this->_useIdentityMap = $data[10];
     }
 
+
```
Comment (thePanz) [inline lib/Doctrine/Table.php:3055]: Remove the extra line here

File: tests/Ticket/1783TestCase.php [hunk 7]
```diff
@@ -12,9 +12,12 @@ public function testValidateLargeIntegers()
         $this->manager->setAttribute(Doctrine_Core::ATTR_VALIDATE, Doctrine_Core::VALIDATE_ALL);        
 
         $test = new Ticket_1783();
+
         $test->bigint = PHP_INT_MAX + 1;
-        
-        $this->assertTrue($test->isValid());
+
+        // This test works on php 32bit version because float allow to represent bigger value than a int.
```
Comment (thePanz) [inline tests/Ticket/1783TestCase.php:18]: Good catch! :+1:

File: lib/Doctrine/Table.php [hunk 8]
```diff
@@ -1132,6 +1132,11 @@ public function processOrderBy($alias, $orderBy, $columnNames = false)
            $alias = $this->getComponentName();
         }
 
+        // Php8.1 require a string
+        if(null === $orderBy) {
```
Comment (thirsch) [inline lib/Doctrine/Table.php:1136]:
```suggestion
        if (null === $orderBy) {
```

File: lib/Doctrine/Task.php [hunk 9]
```diff
@@ -53,7 +53,7 @@ public function __construct($dispatcher = null)
     {
         $this->dispatcher = $dispatcher;
 
-        $taskName = $this->getTaskName();
+        $taskName = (string)$this->getTaskName();
```
Comment (thirsch) [inline lib/Doctrine/Task.php:56]:
```suggestion
        $taskName = (string) $this->getTaskName();
```

File: lib/Doctrine/Validator.php [hunk 10]
```diff
@@ -126,9 +126,9 @@ public static function validateLength($value, $type, $maximumLength)
     public static function getStringLength($string)
     {
         if (function_exists('mb_strlen')) {
-            return mb_strlen($string, 'utf8');
+            return mb_strlen((string)$string, 'utf8');
```
Comment (thirsch) [inline lib/Doctrine/Validator.php:129]:
```suggestion
            return mb_strlen((string) $string, 'utf8');
```

File: lib/Doctrine/Validator.php [hunk 11]
```diff
@@ -126,9 +126,9 @@ public static function validateLength($value, $type, $maximumLength)
...
         } else {
-            return strlen(utf8_decode($string));
+            return strlen(utf8_decode((string)$string));
```
Comment (thirsch) [inline lib/Doctrine/Validator.php:131]:
```suggestion
            return strlen(utf8_decode((string) $string));
```

File: lib/Doctrine/Validator/Notblank.php [hunk 12]
```diff
@@ -41,6 +41,6 @@ class Doctrine_Validator_Notblank extends Doctrine_Validator_Driver
      */
     public function validate($value)
     {
-        return (trim($value) !== '' && $value !== null);
+        return (trim((string) $value) !== '' && $value !== null);
```
Comment (thirsch) [inline lib/Doctrine/Validator/Notblank.php:44]: You could either swap the two operands and remove the cast or remove the second condition.

**Review by thirsch — CHANGES_REQUESTED** [review]

File: lib/Doctrine/Record.php (hunk 5 above)
Comment (Tybaze) [reply lib/Doctrine/Record.php]: Fixed !

File: lib/Doctrine/Table.php (hunk 6 above)
Comment (Tybaze) [reply lib/Doctrine/Table.php]: Fixed, and also reviewed phpdoc which trigger some errors in my IDE.
Thank you !

File: lib/Doctrine/Validator/Notblank.php (hunk 12 above)
Comment (Tybaze) [reply lib/Doctrine/Validator/Notblank.php]: Thank you very much, it avoid a useless cast.

File: lib/Doctrine/Table.php [hunk 13]
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
...
+        $this->hasDefaultValues = $data[7];
+        $this->_options = $data[8];
+        $this->_invokedMethods = $data[9];
+        $this->_useIdentityMap = $data[10];
     }
 
+    /**
+     * Table loaded from cache, instead of __construct()
+     *
+     * @param Doctrine_Connection $conn
```
Comment (thePanz) [inline lib/Doctrine/Table.php:3056]: Removing, as it does not bring more clarity :)

```suggestion
```

**Review by thePanz — CHANGES_REQUESTED** [review]

File: lib/Doctrine/Connection/Sqlite.php (hunk 4 above)
Comment (alquerci) [reply lib/Doctrine/Connection/Sqlite.php]: Sadly, it is the PHP Doc that is wrong here.

The interface to respect is defined on [Doctrine_Connection::connect()](https://github.com/FriendsOfSymfony1/doctrine1/blob/master/lib/Doctrine/Connection.php#L455)

File: lib/Doctrine/Connection/Sqlite.php [hunk 14]
```diff
@@ -84,11 +90,21 @@ public function __construct(Doctrine_Manager $manager, $adapter)
     public function connect() 
     {
         if ($this->isConnected) {
-            return false;
+            return;
```
Comment (alquerci) [inline lib/Doctrine/Connection/Sqlite.php:93]: Needs to respect the interface defined on [Doctrine_Connection::connect()](https://github.com/FriendsOfSymfony1/doctrine1/blob/master/lib/Doctrine/Connection.php#L455)

The PHP Doc is wrong here.

File: tests/Ticket/1783TestCase.php [hunk 15]
```diff
@@ -12,9 +12,12 @@ public function testValidateLargeIntegers()
...
+        // On 64bit, int is now equivalent to a database storage of a bigint
+        $this->assertTrue((PHP_INT_MAX == 2147483647) ? $test->isValid() : true);
```
Comment (alquerci) [inline tests/Ticket/1783TestCase.php:20]: Yes, nice catch, it is a good track, but the conclusion is wrong.

The table definition will correspond to MySql type `BIGINT UNSIGNED`.

Which has a maximum value of `2^64 - 1`.
However, `PHP_INT_MAX` on 64bit system architecture is `2^63-1` as PHP integer is not unsigned.

Using `getErrorStackAsString()` we can see what validator failed.
We can see that the error comes from `unsigned` validation.

* More over, there is no validation on the length for integer.
    https://github.com/FriendsOfSymfony1/doctrine1/blob/master/lib/Doctrine/Validator.php#L95-L96
* Digging deeper of the `Unsigned` validator, you can see that the regex is in fault.
    It depends on how the value is converted to string, when it is bigger than PHP_INT_MAX it contains `E` the exponential symbol.

**Conclusion**
The issue still exists for 64bit system.

For now, we can ignore this test.
https://github.com/FriendsOfSymfony1/doctrine1/pull/56/commits/6a08c161467bddf89afd0eeeacfa6abf085d2c47

File: lib/Doctrine/Task.php [hunk 16]
```diff
@@ -53,7 +53,7 @@ public function __construct($dispatcher = null)
     {
         $this->dispatcher = $dispatcher;
 
-        $taskName = $this->getTaskName();
+        $taskName = (string) $this->getTaskName();
```
[the remark this answers is not in the document]
Comment (alquerci) [reply lib/Doctrine/Task.php:56]: There is no test case when task name is an empty string.

So the minimum things to do should be
* Handle `null` as @thePanz said
* Add test case for empty string, it seems an extreme edge case, but why not,

File: lib/Doctrine/Connection/Statement.php [hunk 17]
```diff
@@ -317,7 +317,7 @@ public function execute($params = array())
      */
     public function fetch($fetchMode = Doctrine_Core::FETCH_BOTH,
                           $cursorOrientation = Doctrine_Core::FETCH_ORI_NEXT,
-                          $cursorOffset = null)
+                          $cursorOffset = 0)
```
Comment (alquerci) [inline lib/Doctrine/Connection/Statement.php:320]: As it is, `null` on the interface, this is a violation.

Can you try to fix without this violation?

File: lib/Doctrine/Table.php [hunk 18]
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
...
+        $this->_columnNames = $data[5];
+        $this->columnCount = $data[6];
+        $this->hasDefaultValues = $data[7];
+        $this->_options = $data[8];
+        $this->_invokedMethods = $data[9];
+        $this->_useIdentityMap = $data[10];
     }
 
+    /**
+     * Table loaded from cache, instead of __construct()
```
Comment (alquerci) [inline lib/Doctrine/Table.php:3054]: The method name is a bit misleading.

What does it do ?

```suggestion
     * Creates new instance and initialize it from cache.
```

File: lib/Doctrine/Table.php [hunk 19]
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
...
+     *
+     * @param Doctrine_Connection $conn
+     * @return void
```
Comment (alquerci) [inline lib/Doctrine/Table.php:3057]:
```suggestion
```

File: lib/Doctrine/Table.php [hunk 20]
```diff
@@ -2991,26 +3027,35 @@ public function serialize()
...
+        );
     }
 
-    public function unserialize($data)
-    {
-        $all = unserialize($data);
+    /**
+     * Unserializes a Doctrine_Record instance for php 7.4+
+     *
+     * @param array $serialized
```
Comment (alquerci) [inline lib/Doctrine/Table.php:3036]:
```suggestion
     * @param array $data
```

**Review by alquerci — CHANGES_REQUESTED** [review]

**Comment by alquerci** [comment]
A pass of code structure fixes will be appreciated.

File: lib/Doctrine/Connection/Sqlite.php [hunk 21]
```diff
@@ -84,11 +90,21 @@ public function __construct(Doctrine_Manager $manager, $adapter)
...
+        // If customer configure it
+        $hasConfigureStringify = (isset($this->pendingAttributes[Doctrine_Core::ATTR_STRINGIFY_FETCHES]));
+
         parent::connect();
 
+        if(!$hasConfigureStringify) {
+            // PHP8.1 require default to true to keep BC
+            // https://www.php.net/manual/en/migration81.incompatible.php#migration81.incompatible.pdo.sqlite
+            // Can be overwritten by user later
+            $this->dbh->setAttribute(PDO::ATTR_STRINGIFY_FETCHES, true);
```
[the remark this answers is not in the document]
Comment (Tybaze) [reply lib/Doctrine/Connection/Sqlite.php:105]: In the case if the user modify dbh attribute with PDO::ATTR_STRINGIFY_FETCHES, he do not respect doctrine1 behavior, it will be overwritten.

If the user set Doctrine_Core::ATTR_STRINGIFY_FETCHES to false, the value will still be false, and the new php 8.1 behavior will apply.

Doctrine3 decided to drop the handling, delegating the fault to the pdo driver. This is normal for an evolutive framework to move to the new behavior
https://github.com/doctrine/dbal/pull/5229/files

As doctrine1 target is to keep max BC, it's better to do our max to keep the previous behavior

File: lib/Doctrine/Task.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (Tybaze) [reply lib/Doctrine/Task.php:56]: Fixed by handling the null value : 711ec88b83e0a9ab64d1c7765206e220cbfca85a

I miss some time to create the unit test about this rare case

File: lib/Doctrine/Task.php [hunk 22]
```diff
@@ -56,7 +56,7 @@ public function __construct($dispatcher = null)
         $taskName = (string) $this->getTaskName();
 
         //Derive the task name only if it wasn't entered at design-time
-        if (! strlen($taskName)) {
+        if (null === $taskName || ! strlen($taskName)) {
```
Comment (thePanz) [inline lib/Doctrine/Task.php:59]:
```suggestion
        if (null === $taskName || 0 !== strlen($taskName)) {
```

**Review by thePanz — CHANGES_REQUESTED** [review]

File: lib/Doctrine/Connection/Sqlite.php (hunk 4 above)
Comment (Tybaze) [reply lib/Doctrine/Connection/Sqlite.php:93]: 908bcbb316aa372ed887a4973121af5b8d890ee7 I have set a "@return boolean", exactly like the abstract Doctrine_Connection::connect()

is it ok ?

File: lib/Doctrine/Task.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (alquerci) [reply lib/Doctrine/Task.php:56]: > I miss some time to create the unit test about this rare case

How it is or will be going ?

This try to add a proof that it work as it should, is like important for the ability to maintain doctrine1 code.

If you have time, I will be interested in your return experience.

---

Between I have a doubt on the last commit https://github.com/FriendsOfSymfony1/doctrine1/commit/86519ea755905c2004932fd3b40b321e1d0543e2 behaviour.

File: lib/Doctrine/Task.php (hunk 22 above)
Comment (alquerci) [reply lib/Doctrine/Task.php]: @thePanz I do not understand this suggestion. It changes the behaviour or the structure ?

Comment (thePanz) [reply lib/Doctrine/Task.php]: imho it makes more explicit the strlen evaluation, and the fact that strlen returns an integer (not a boolean)

File: lib/Doctrine/Connection/Sqlite.php (hunk 4 above)
Comment (thirsch) [reply lib/Doctrine/Connection/Sqlite.php:93]: Hi @thePanz, you've given a thumbs up to the change from @Tybaze therefore it should be ok to accept this pr, shouldn't it?

File: lib/Doctrine/Task.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (alquerci) [reply lib/Doctrine/Task.php:56]: @thirsch It left this issue.

Currently, custom task name is ignored when set on `taskName` property.

```php
if (null === $taskName || 0 !== strlen($taskName)) {
```

Should be at least

```php
if (null === $taskName || 0 === strlen($taskName)) {
```

[the remark this answers is not in the document]
Comment (Tybaze) [reply lib/Doctrine/Task.php:56]: Just fixed in 09da1cd899410b2a026fae178ec57440baa62e2a

**Comment by Tybaze** [comment]
Hi Everyone ^^
Do I miss something to merge this PR ?

File: lib/Doctrine/Task.php (hunk 22 above)
Comment (Tybaze) [reply lib/Doctrine/Task.php]: Fixed on 09da1cd899410b2a026fae178ec57440baa62e2a

File: lib/Doctrine/Task.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (thePanz) [reply lib/Doctrine/Task.php:56]: @Tybaze if we cast to string `(string) $this->getTaskName();` , then the `if (null === $taskName` will always be false.

With the casting we can simplify the condition to

```php
$taskName = (string) $this->getTaskName();
if ('' === trim($taskName)) {
    $taskName = self::deriveTaskName(get_class($this));
}
```

What do you think?

[the remark this answers is not in the document]
Comment (thePanz) [reply lib/Doctrine/Task.php:56]: This covers the weird case of a task with a set of spaces as name

**Review by thePanz — CHANGES_REQUESTED** [review]

File: lib/Doctrine/Task.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (alquerci) [reply lib/Doctrine/Task.php:56]: @thePanz you are right.

There are many ways to make this work.

But the main issue here is that there no test for that case.

> How can you prevent harm to the structure of your code if you don’t have the tests that allow you to clean it?
--- "Clean Craftsmanship: Disciplines, Standards, and Ethics (Robert C. Martin Series) (English Edition)" write by Robert C. Martin

---

Let me try to add a repeatable proof that every element of the code works as it should.

**Comment by alquerci** [comment]
@thePanz I remind you that this PR required following ones

* FriendsOfSymfony1/doctrine1#56
* FriendsOfSymfony1/doctrine1#86

File: lib/Doctrine/Task.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (alquerci) [reply lib/Doctrine/Task.php:56]: Here we go

https://github.com/Tybaze/doctrine1/pull/2

**Comment by Tybaze** [comment]
Rebase:
Merge commits about the TaskName declaration, using the latest @thePanz version:
`if ('' === trim($taskName)) {`

Merge commits review of this PR about annotation, and conding style.

**Comment by thePanz** [comment]
@Tybaze coudl you rebase on the latest `master` branch?
- the first two commits from xNatek are already merged
- squash commits to group related changes togerter
- remove the "merge master" commits from my side (I tried to use the GitHub feature, but we should not have those commits in the project's history)

Thanks! then we will be able to merge this! :tada:

**Comment by Tybaze** [comment]
> @Tybaze coudl you rebase on the latest `master` branch?
> 
>     * the first two commits from xNatek are already merged
> 
>     * squash commits to group related changes togerter
> 
>     * remove the "merge master" commits from my side (I tried to use the GitHub feature, but we should not have those commits in the project's history)
> 
> 
> Thanks! then we will be able to merge this! 🎉

- Rebased on master (including xNatek commit)
- Squashed by modification type (Compat 8.0, 8.1, PR Review)
- Cleaned Merge master

@thePanz  You're welcome ! 
Sorry for the delay ^^

**Comment by thePanz** [comment]
Merged @Tybaze @alquerci 
Thanks for your efforts!
